File: exam-conductor/archiveDCR/libs/exampen-proto/typescript/src/page.ts
/** Page and copy-upload types: page images, miss indicators, copy pages. */

import type {
  AuthoritativeSource,
  CopyAuthoritativeCandidate,
  MissIndicatorState,
} from "./enums";

/** NATS event: rendered page image ready for AI pipeline. */
export interface PageReadyEvent {
  event_id: string;
  event_type: "page.ready";
  event_version: "1.0.0";
  occurred_at: string;
  exam_id: string;
  student_id: string;
  page_id: string;
  page_number: number;
  image_uri: string;
  vector_uri?: string;
  authoritative_source: AuthoritativeSource;
  question_ids?: string[];
}

/** NATS event: photographed copy page ingested. */
export interface CopyReadyEvent {
  event_id: string;
  event_type: "copy.ready";
  event_version: "1.0.0";
  occurred_at: string;
  exam_id: string;
  student_id: string;
  page_number: number;
  copy_image_uri: string;
  authoritative_candidate?: CopyAuthoritativeCandidate;
}

/** Result of uploading a photographed answer page. */
export interface CopyUploadResult {
  exam_id: string;
  student_id: string;
  page_number: number;
  copy_image_uri: string;
  data_source: "copy_image";
}

/** A single copy page record. */
export interface CopyPage {
  page_number: number;
  copy_image_uri: string;
  authoritative_source?: AuthoritativeSource;
}

/** Single cell in the miss indicator matrix. */
export interface MissIndicatorCell {
  student_id: string;
  question_id: string;
  state: MissIndicatorState;
}

/** Student-by-question miss indicator matrix for a teacher view. */
export interface MissIndicatorMatrix {
  exam_id: string;
  students: string[];
  questions: string[];
  cells: MissIndicatorCell[];
}

File: exam-conductor/archiveDCR/libs/exampen-proto/python/exampen_proto/page.py
"""Page and copy-upload models: page images, miss indicators, copy pages."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from .enums import AuthoritativeSource, CopyAuthoritativeCandidate, MissIndicatorState


class PageReadyEvent(BaseModel):
    """NATS event: rendered page image ready for AI pipeline."""

    event_id: str
    event_type: str = "page.ready"
    event_version: str = "1.0.0"
    occurred_at: datetime
    exam_id: UUID
    student_id: str
    page_id: str
    page_number: int
    image_uri: str
    vector_uri: Optional[str] = None
    authoritative_source: AuthoritativeSource
    question_ids: Optional[list[str]] = None


class CopyReadyEvent(BaseModel):
    """NATS event: photographed copy page ingested."""

    event_id: str
    event_type: str = "copy.ready"
    event_version: str = "1.0.0"
    occurred_at: datetime
    exam_id: UUID
    student_id: str
    page_number: int
    copy_image_uri: str
    authoritative_candidate: Optional[CopyAuthoritativeCandidate] = None


class CopyUploadResult(BaseModel):
    """Result of uploading a photographed answer page."""

    exam_id: UUID
    student_id: str
    page_number: int
    copy_image_uri: str
    data_source: str = "copy_image"


class CopyPage(BaseModel):
    """A single copy page record."""

    page_number: int
    copy_image_uri: str
    authoritative_source: Optional[AuthoritativeSource] = None


class MissIndicatorCell(BaseModel):
    """Single cell in the miss indicator matrix."""

    student_id: str
    question_id: str
    state: MissIndicatorState


class MissIndicatorMatrix(BaseModel):
    """Student-by-question miss indicator matrix for a teacher view."""

    exam_id: UUID
    students: list[str]
    questions: list[str]
    cells: list[MissIndicatorCell]
